db: hold commit pipeline mutex when closing DB

*This was the result of some conversation on on #1719.*

The commit pipeline mutex is used as an external synchronization mechanism for log writes. If an operation (set, delete, etc.) is performed on the DB after the DB has been closed, the operation can cause a data race in the log writer. This was observed in cockroachdb/cockroach#81511, following a change that allowed concurrent operations post-close (see the original change in cockroachdb/cockroach#81133).

The data race is most likely caused by the omission of a commit-pipeline lock in `DB.Close`. Closing the WAL writer is required to hold the commit pipeline mutex normally (eg, `makeRoomForWrite` always holds `d.commit.mu`).

We could also consider adjusting `DB.Close` to make concurrent operations a little easier to catch. For example, changing the first line of `Close()` to swap `d.closed`'s value to `ErrClosed`, increasing the chances a concurrent operation observes the `ErrClosed`.

While the data race has since been fixed above Pebble (see cockroachdb/cockroach#81813), and despite the fact that calling a method on the DB post-close is undefined, we should fix the issue in Pebble.

The following is a reliable reproducer:

```go
package pebble

import (
	"sync"
	"testing"

	"github.com/cockroachdb/pebble/vfs"
	"github.com/stretchr/testify/require"
)

func Test(t *testing.T) {
	db, err := Open("", &Options{
		FS: vfs.NewMem(),
	})
	require.NoError(t, err)

	// Two goroutines racing each other. One attempts to Set a record into the
	// DB, the other attempts to close the DB.
	var wg sync.WaitGroup

	wg.Add(1)
	go func() {
		defer func() {
			if r := recover(); r != nil {
				t.Logf("WARN: saw panic on record write: %s", r)
			}
			wg.Done()
		}()

		t.Logf("writing record ...")
		err := db.Set([]byte("foo"), nil, nil)
		if err != nil {
			t.Logf("WARN: saw error on record write: %s", err)
		}
	}()

	wg.Add(1)
	go func() {
		defer wg.Done()
		t.Logf("closing db ...")
		err := db.Close()
		require.NoError(t, err)
	}()

	wg.Wait()
	t.Logf("done")
}
```

```bash
$ go test -run '^Test$' -race -count 1 .
```

<details>
<summary>Output</summary>

```console
==================
WARNING: DATA RACE
Read at 0x00c0002ae000 by goroutine 27:
  github.com/cockroachdb/pebble/record.(*LogWriter).emitEOFTrailer()
      /Users/nickt/Development/pebble/record/log_writer.go:678 +0x4b
  github.com/cockroachdb/pebble/record.(*LogWriter).Close()
      /Users/nickt/Development/pebble/record/log_writer.go:589 +0x32
  github.com/cockroachdb/pebble.(*DB).Close()
      /Users/nickt/Development/pebble/db.go:1203 +0x64e
  github.com/cockroachdb/pebble.Test.func2()
      /Users/nickt/Development/pebble/test_test.go:41 +0xc8

Previous write at 0x00c0002ae000 by goroutine 26:
  sync/atomic.StoreInt32()
      /nix/store/0bh8m4vqb7c7n4sn73ka53yn45yj3zd0-go-1.18.2/share/go/src/runtime/race_amd64.s:229 +0xb
  sync/atomic.StoreInt32()
      <autogenerated>:1 +0x1a
  github.com/cockroachdb/pebble/record.(*LogWriter).SyncRecord()
      /Users/nickt/Development/pebble/record/log_writer.go:646 +0xe4
  github.com/cockroachdb/pebble.(*DB).commitWrite()
      /Users/nickt/Development/pebble/db.go:851 +0x51a
  github.com/cockroachdb/pebble.(*DB).commitWrite-fm()
      <autogenerated>:1 +0x64
  github.com/cockroachdb/pebble.(*commitPipeline).prepare()
      /Users/nickt/Development/pebble/commit.go:379 +0x2db
  github.com/cockroachdb/pebble.(*commitPipeline).Commit()
      /Users/nickt/Development/pebble/commit.go:253 +0xb7
  github.com/cockroachdb/pebble.(*DB).Apply()
      /Users/nickt/Development/pebble/db.go:757 +0x465
  github.com/cockroachdb/pebble.(*DB).Set()
      /Users/nickt/Development/pebble/db.go:575 +0x108
  github.com/cockroachdb/pebble.Test.func1()
      /Users/nickt/Development/pebble/test_test.go:31 +0x10c

Goroutine 27 (running) created at:
  github.com/cockroachdb/pebble.Test()
      /Users/nickt/Development/pebble/test_test.go:38 +0x4ee
  testing.tRunner()
      /nix/store/0bh8m4vqb7c7n4sn73ka53yn45yj3zd0-go-1.18.2/share/go/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /nix/store/0bh8m4vqb7c7n4sn73ka53yn45yj3zd0-go-1.18.2/share/go/src/testing/testing.go:1486 +0x47

Goroutine 26 (finished) created at:
  github.com/cockroachdb/pebble.Test()
      /Users/nickt/Development/pebble/test_test.go:22 +0x3f2
  testing.tRunner()
      /nix/store/0bh8m4vqb7c7n4sn73ka53yn45yj3zd0-go-1.18.2/share/go/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /nix/store/0bh8m4vqb7c7n4sn73ka53yn45yj3zd0-go-1.18.2/share/go/src/testing/testing.go:1486 +0x47
==================
--- FAIL: Test (0.00s)
    test_test.go:30: writing record ...
    test_test.go:40: closing db ...
    test_test.go:46: done
    testing.go:1312: race detected during execution of test
FAIL
FAIL    github.com/cockroachdb/pebble   0.867s
FAIL
```
</details>

--- a/compaction.go
+++ b/compaction.go
@@ -1520,11 +1520,11 @@ func (d *DB) maybeScheduleDelayedFlush(tbl *memTable, dur time.Duration) {
 			defer d.mu.Unlock()
 
 			// NB: The timer may fire concurrently with a call to Close.  If a
-			// Close call beat us to acquiring d.mu, d.closed is 1, and it's
-			// too late to flush anything. Otherwise, the Close call will
-			// block on locking d.mu until we've finished scheduling the flush
-			// and set `d.mu.compact.flushing` to true. Close will wait for
-			// the current flush to complete.
+			// Close call beat us to acquiring d.mu, d.closed holds ErrClosed,
+			// and it's too late to flush anything. Otherwise, the Close call
+			// will block on locking d.mu until we've finished scheduling the
+			// flush and set `d.mu.compact.flushing` to true. Close will wait
+			// for the current flush to complete.
 			if d.closed.Load() != nil {
 				return
 			}

--- a/db.go
+++ b/db.go
@@ -1211,6 +1211,18 @@ func (d *DB) NewSnapshot() *Snapshot {
 // or to call Close concurrently with any other DB method. It is not valid
 // to call any of a DB's methods after the DB has been closed.
 func (d *DB) Close() error {
+	// Lock the commit pipeline for the duration of Close. This prevents a race
+	// with makeRoomForWrite. Rotating the WAL in makeRoomForWrite requires
+	// dropping d.mu several times for I/O. If Close only holds d.mu, an
+	// in-progress WAL rotation may re-acquire d.mu only once the database is
+	// closed.
+	//
+	// Additionally, locking the commit pipeline makes it more likely that
+	// (illegal) concurrent writes will observe d.closed.Load() != nil, creating
+	// more understable panics if the database is improperly used concurrently
+	// during Close.
+	d.commit.mu.Lock()
+	defer d.commit.mu.Unlock()
 	d.mu.Lock()
 	defer d.mu.Unlock()
 	if err := d.closed.Load(); err != nil {

--- a/range_del_test.go
+++ b/range_del_test.go
@@ -7,8 +7,10 @@ package pebble
 import (
 	"bytes"
 	"fmt"
+	"runtime"
 	"strconv"
 	"strings"
+	"sync"
 	"testing"
 	"time"
 
@@ -18,6 +20,7 @@ import (
 	"github.com/cockroachdb/pebble/sstable"
 	"github.com/cockroachdb/pebble/vfs"
 	"github.com/stretchr/testify/require"
+	"golang.org/x/exp/rand"
 )
 
 func TestRangeDel(t *testing.T) {
@@ -195,6 +198,59 @@ func TestFlushDelay(t *testing.T) {
 		<-flushed
 	}
 	require.NoError(t, d.Close())
+}
+
+func TestFlushDelayStress(t *testing.T) {
+	rng := rand.New(rand.NewSource(uint64(time.Now().UnixNano())))
+	opts := &Options{
+		FS:                    vfs.NewMem(),
+		Comparer:              testkeys.Comparer,
+		FlushDelayDeleteRange: time.Duration(rng.Intn(10)+1) * time.Millisecond,
+		FlushDelayRangeKey:    time.Duration(rng.Intn(10)+1) * time.Millisecond,
+		FormatMajorVersion:    FormatNewest,
+		MemTableSize:          8192,
+	}
+
+	const runs = 100
+	for run := 0; run < runs; run++ {
+		d, err := Open("", opts)
+		require.NoError(t, err)
+
+		now := time.Now().UnixNano()
+		writers := runtime.GOMAXPROCS(0)
+		var wg sync.WaitGroup
+		wg.Add(writers)
+		for i := 0; i < writers; i++ {
+			rng := rand.New(rand.NewSource(uint64(now) + uint64(i)))
+			go func() {
+				const ops = 100
+				defer wg.Done()
+
+				var k1, k2 [32]byte
+				for j := 0; j < ops; j++ {
+					switch rng.Intn(3) {
+					case 0:
+						randStr(k1[:], rng)
+						randStr(k2[:], rng)
+						require.NoError(t, d.DeleteRange(k1[:], k2[:], nil))
+					case 1:
+						randStr(k1[:], rng)
+						randStr(k2[:], rng)
+						require.NoError(t, d.RangeKeySet(k1[:], k2[:], []byte("@2"), nil, nil))
+					case 2:
+						randStr(k1[:], rng)
+						randStr(k2[:], rng)
+						require.NoError(t, d.Set(k1[:], k2[:], nil))
+					default:
+						panic("unreachable")
+					}
+				}
+			}()
+		}
+		wg.Wait()
+		time.Sleep(time.Duration(rng.Intn(10)+1) * time.Millisecond)
+		require.NoError(t, d.Close())
+	}
 }
 
 // Verify that range tombstones at higher levels do not unintentionally delete

